packages/components/src/composite/legacy/index.tsx

```diff
@@ -0,0 +1,177 @@
+/**
+ * Composite is a component that may contain navigable items represented by
+ * CompositeItem. It's inspired by the WAI-ARIA Composite Role and implements
+ * all the keyboard navigation mechanisms to ensure that there's only one
+ * tab stop for the whole Composite element. This means that it can behave as
+ * a roving tabindex or aria-activedescendant container.
+ *
+ * @see https://ariakit.org/components/composite
+ */
+
+/**
+ * WordPress dependencies
+ */
+import {
+	forwardRef,
+	useId as useGeneratedId,
+	useRef,
+	useState,
+} from '@wordpress/element';
+import deprecated from '@wordpress/deprecated';
+
+/**
+ * Internal dependencies
+ */
+import * as Current from '../current';
+
+type Component = ( ...any: any[] ) => any;
+
+type CompositeStore = ReturnType< typeof Current.useCompositeStore >;
+type CompositeStoreState = { store: CompositeStore };
+type CompositeStoreProps = NonNullable<
+	Parameters< typeof Current.useCompositeStore >[ 0 ]
+>;
+
+type Orientation = 'horizontal' | 'vertical';
+
+export interface InitialState {
+	baseId?: string;
+	unstable_virtual?: boolean;
+	rtl?: boolean;
+	orientation?: Orientation;
+	currentId?: string;
+	loop?: boolean | Orientation;
+	wrap?: boolean | Orientation;
+	shift?: boolean;
+}
+
+type BaseIdState = Required< Pick< InitialState, 'baseId' > >;
+type BaseId = BaseIdState[ 'baseId' ];
+type MappedInitialState = CompositeStoreProps & BaseIdState;
+
+export type CompositeState = CompositeStoreState & BaseIdState;
+export type CompositeStateProps =
+	| { state: CompositeState }
+	| ( CompositeState & { state?: never } );
+export type CompositeProps< C extends Component > = CompositeStateProps &
+	Parameters< C >[ 0 ];
+
+type ManagedProps = CompositeStoreState & {
+	props: Record< PropertyKey, any >;
+};
+
+function showDeprecationMessage( ...props: Parameters< typeof deprecated > ) {
+	if ( 'test' !== process.env.NODE_ENV ) {
+		deprecated( ...props );
+	}
+}
+
+const idMap = new Map< BaseId, React.MutableRefObject< number > >();
+
+function useBaseId( preferredBaseId?: BaseId ): BaseId {
+	const generatedId = useGeneratedId();
+	const [ baseId ] = useState( preferredBaseId || generatedId );
+	idMap.set( baseId, useRef( 0 ) );
+
+	return baseId;
+}
+
+function useId( baseId: BaseId, preferredId?: string ) {
+	const counter = idMap.get( baseId ) ?? { current: 0 };
+	const [ id ] = useState( () => {
+		if ( preferredId ) return preferredId;
+		return `${ baseId }-${ ++counter.current }`;
+	} );
+
+	return id;
+}
```

Comment:
This was being done to provide the same consistency that the old composite functionality offered – `useInstanceId` doesn't keep the same ID for the same component on separate renders. However, this probably isn't necessary, and if a consumer really needs that behaviour, it can be achieved by providing their own `baseId`. Switched it.